### PR TITLE
provider/aws: AWS SpotFleet Requests now works with Subnets and AZs

### LINKED ISSUE #8150: aws_spot_fleet_request crashes when launch_specification contains only subnet_id and not availability_zone
Hi,

Given the following:

```
resource "aws_spot_fleet_request" "packer" {
    lifecycle {
        create_before_destroy = true
    }

    iam_fleet_role = "${aws_iam_role.spot_fleet.arn}"
    target_capacity = 1
    spot_price = "0.001"
    valid_until = "2017-01-31T17:00:00Z"
    terminate_instances_with_expiration = false

    launch_specification {
        instance_type = "c3.large"
        spot_price = "0.105"
        ami = "ami-a52ab4b2"
        subnet_id = "${element(data.terraform_remote_state.vpc.private_subnet_ids, 0)}"
        key_name = "nw-prod"
        vpc_security_group_ids = ["${aws_security_group.ecs_sg.id}"]
    }
}
```

Results in a crash during application, but the fleet is actually created.

```
* aws_spot_fleet_request.packer: unexpected EOF
2016/08/11 15:17:29 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:
```

By simply adding `availability_zone = "${element(data.terraform_remote_state.vpc.azs, 0)}"` then the apply succeeds.

Per the docs we should be able to supply just a subnet ID: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet-examples.html#fleet-config2

PS.  It would GREAT if `launch_specification` supported count so we can map it to our subnets.

### LINKED ISSUE #8203: Spot fleet request failed with compatible availability zone and subnet
With Terraform v0.7.0, we got following error messages when request spot fleet.

* aws_spot_fleet_request.chunkserver.1: Error requesting spot fleet: [WARN] Error creating Spot fleet request, retrying: InvalidSpotFleetRequestConfig: The specified Subnet: subnet-XYZ cannot be used with the specified Availability Zone: us-west-2a.

While subnet-XYZ and us-west-2a is our default settings when we use WebUI and command tools to request spot fleet and they worked. But it failed with Terraform. TF_LOG=TRACE also shows the request string contains the correct availability zone and subnet id.

### PR DESCRIPTION
**Pull request by stack72**
Fixes #8150 
Fixes #8203 

Also added a complete set of tests that reflect all of the use cases
that Amazon document
http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet-examples.html

It is important to note there that Terraform will be suggesting that
users create multiple launch configurations rather than AWS's version of
combing values into CSV based parameters. This will ensure that we are
able to enforce the correct state

Also note that `associate_public_ip_address` now defaults to `false` - a migration has been
included in this PR to migration users of this functionality. This needs
to be noted in the changelog. The last part of changing functionality
here is waiting for the state of the request to become `active`. Before
we get to this state, we cannot guarantee that Amazon have accepted the
request or it could have failed validation.

```
% make testacc TEST=./builtin/providers/aws
% TESTARGS='-run=TestAccAWSSpotFleetRequest_'
% 2 ↵
==> Checking that code complies with gofmt requirements...
/Users/stacko/Code/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/22 15:44:21 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v
-run=TestAccAWSSpotFleetRequest_ -timeout 120m
=== RUN   TestAccAWSSpotFleetRequest_changePriceForcesNewRequest
--- PASS: TestAccAWSSpotFleetRequest_changePriceForcesNewRequest (133.90s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion (76.67s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList (75.22s)
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList (96.95s)
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz (74.44s)
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet (97.82s)
=== RUN   TestAccAWSSpotFleetRequest_overriddingSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_overriddingSpotPrice (76.22s)
=== RUN   TestAccAWSSpotFleetRequest_diversifiedAllocation
--- PASS: TestAccAWSSpotFleetRequest_diversifiedAllocation (79.81s)
=== RUN   TestAccAWSSpotFleetRequest_withWeightedCapacity
--- PASS: TestAccAWSSpotFleetRequest_withWeightedCapacity (77.15s)
=== RUN   TestAccAWSSpotFleetRequest_CannotUseEmptyKeyName
--- PASS: TestAccAWSSpotFleetRequest_CannotUseEmptyKeyName (0.00s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    788.184s
```

### CODE REVIEW

**Comment by kwilczynski** [comment]
This is some amazing work!

File: builtin/providers/aws/resource_aws_spot_fleet_request.go [hunk 1]
```diff
@@ -49,7 +52,7 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
 						"associate_public_ip_address": &schema.Schema{
 							Type:     schema.TypeBool,
 							Optional: true,
-							Default:  true,
+							Default:  false,
```
Comment (jen20) [inline builtin/providers/aws/resource_aws_spot_fleet_request.go]: :+1: probably will need to document this as breaking?

File: builtin/providers/aws/resource_aws_spot_fleet_request.go [hunk 2]
```diff
@@ -773,7 +797,7 @@ func launchSpecToMap(
 	}
 
 	if l.WeightedCapacity != nil {
-		m["weighted_capacity"] = fmt.Sprintf("%.3f", aws.Float64Value(l.WeightedCapacity))
+		m["weighted_capacity"] = floatToString(*l.WeightedCapacity)
```
Comment (jen20) [inline builtin/providers/aws/resource_aws_spot_fleet_request.go]: Any reason for the helper function here when it's a one-liner from sodlib?

**Comment by jen20** [comment]
Minor question but LGTM other than that. Certainly makes this resource a lot more useful!

**Comment by stack72** [comment]
@jen20 you are correct - no need for a 1 line helper :) Clutter --

**Pull request merged**